core/indexer/process/transactions/transactionsProcessor.go

```diff
@@ -0,0 +1,239 @@
+		if !ok {
+			continue
+		}
+
+		tx = tdp.addScResultInfoInTx(dbScResult, tx)
+		countScResults[string(scResult.OriginalTxHash)]++
+		delete(scResults, scHash)
+
+		// append child smart contract results
+		scrs := findAllChildScrResults(scHash, scResults)
+		for childScHash, sc := range scrs {
+			childDBScResult := tdp.txBuilder.convertScResultInDatabaseScr(childScHash, sc, header)
+
+			tx = tdp.addScResultInfoInTx(childDBScResult, tx)
+			countScResults[string(scResult.OriginalTxHash)]++
+		}
+	}
+
+	return dbSCResults, countScResults
+}
+
+func findAllChildScrResults(hash string, scrs map[string]*smartContractResult.SmartContractResult) map[string]*smartContractResult.SmartContractResult {
+	scrResults := make(map[string]*smartContractResult.SmartContractResult)
+	for scrHash, scr := range scrs {
+		if string(scr.OriginalTxHash) == hash {
+			scrResults[scrHash] = scr
+			delete(scrs, scrHash)
+		}
+	}
+
+	return scrResults
+}
+
+func (tdp *txDatabaseProcessor) addTxsLogsIfNeeded(txs map[string]*types.Transaction) {
+	defer tdp.txLogsProcessor.Clean()
+
+	if !tdp.saveTxsLogsEnabled {
+		return
+	}
+
+	for hash, tx := range txs {
+		txLog, ok := tdp.txLogsProcessor.GetLogFromCache([]byte(hash))
+		if !ok {
+			continue
+		}
+
+		tx.Logs = tdp.prepareTxLog(txLog)
+	}
+}
+
+func (tdp *txDatabaseProcessor) addScResultInfoInTx(dbScResult *types.ScResult, tx *types.Transaction) *types.Transaction {
+	tx.SmartContractResults = append(tx.SmartContractResults, dbScResult)
+
+	if isSCRForSenderWithRefund(dbScResult, tx) {
+		refundValue := stringValueToBigInt(dbScResult.Value)
+		gasUsed, fee := tdp.txFeeCalculator.ComputeGasUsedAndFeeBasedOnRefundValue(tx, refundValue)
+		tx.GasUsed = gasUsed
+		tx.Fee = fee.String()
+	}
+
+	return tx
+}
+
+func (tdp *txDatabaseProcessor) prepareTxLog(log data.LogHandler) *types.TxLog {
+	scAddr := tdp.txBuilder.addressPubkeyConverter.Encode(log.GetAddress())
+	events := log.GetLogEvents()
+
+	txLogEvents := make([]types.Event, len(events))
+	for i, event := range events {
+		txLogEvents[i].Address = hex.EncodeToString(event.GetAddress())
+		txLogEvents[i].Data = hex.EncodeToString(event.GetData())
+		txLogEvents[i].Identifier = hex.EncodeToString(event.GetIdentifier())
+
+		topics := event.GetTopics()
+		txLogEvents[i].Topics = make([]string, len(topics))
+		for j, topic := range topics {
+			txLogEvents[i].Topics[j] = hex.EncodeToString(topic)
+		}
+	}
+
+	return &types.TxLog{
+		Address: scAddr,
+		Events:  txLogEvents,
+	}
+}
+
+func (tdp *txDatabaseProcessor) setTransactionSearchOrder(transactions map[string]*types.Transaction) map[string]*types.Transaction {
+	currentOrder := uint32(0)
+	for _, tx := range transactions {
+		tx.SearchOrder = currentOrder
+		currentOrder++
+	}
+
+	return transactions
+}
+
+// SetTxLogsProcessor -
+func (tdp *txDatabaseProcessor) SetTxLogsProcessor(txLogProcessor process.TransactionLogProcessorDatabase) {
+	tdp.txLogsProcessor = txLogProcessor
```

Comment:
This function is exported. Concurrency issues can happen here. Please refactor.

Reply:
added a mutex.